### ISSUE TITLE
Transform + Generic

### ISSUE THREAD

**Issue by wasd171** [sinclairzx81/typebox#572]
Hello!

I am not sure whether currently type transforms + generic types are working as intended:
```
const Nullable = <T extends TSchema>(t: T) => Type.Union([t, Type.Null()])
const StringToDate = Type.Transform(Type.String())
  .Decode(s => new Date(s))
  .Encode(d => d.toISOString())

const test1 = Value.Decode(StringToDate, '2021-01-01') // <- Date | null, correct
const test2 = Value.Decode(Nullable(StringToDate), '2021-01-01') // <- string | null, incorrect
```

Our type checks pass for `0.31.5` but fail for `0.31.9`

Potentially related to https://github.com/sinclairzx81/typebox/issues/561

**Comment by sinclairzx81** [comment]
@wasd171 Hi!

Hey thanks for reporting. Have just published a revision to resolve this on 0.31.10. There was quite a bit of optimization work done on Decode recently (and some fixes) to try and get this inference into shape, and it appears I missed reimplementing Tuple, Union and Intersect decoding (my bad). Can review the updates at the link below.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUGUDGAFhiAIZwC+cAZlBCHAOQACAzsAHb4A2pwUAehjoMAIwgAPRgFgAUKEixEcAGqluAVwyUadBi3Zde-ISPESBAN3VaZsufggdW8AHIbuvUd20BeOAA8yHAYEjAYHAAmrHhEJKQAfAAUMABcKACUcL4JKCIAdACqHMBOSQDaMDhomPnunkkZALoZDk4ucLgwUJwA5sgQACKk4dl5tchQpM7U0CBJNRj5XT0cvY2tsnBw+YMYjpEYSTE5cBwYAO5ww+HHGZvb+QCiXBCHSZHZuZH5MBAAkrgAPIrPobORtZzwcIuACMYzUmiWewOR1BawGN2wTAATAAGHGwgC0eOJpMYWQEAkCROuI20AB8zh5uDhHFAoPsYJCOjCYDiETZkfs3kd6l4fEl0f0hvSMjhGPjCSSybCKXAqTS4C5Vr04EyOCycJx2Zz8Nz7Ja5EA)

```typescript
import { Type, TSchema } from '@sinclair/typebox'
import { Value } from '@sinclair/typebox/value'

const Nullable = <T extends TSchema>(t: T) => Type.Union([t, Type.Null()])
const StringToDate = Type.Transform(Type.String())
  .Decode(s => new Date(s))
  .Encode(d => d.toISOString())

const test1 = Value.Decode(StringToDate, '2021-01-01')           // <- Date
const test2 = Value.Decode(Nullable(StringToDate), '2021-01-01') // <- Date | null
```
Let me know if this works for you. 
Cheers
S

**Comment by wasd171** [comment]
Hi @sinclairzx81 

Thanks for the quick fix! With `0.31.10` our checks pass again :)
